crates/matrix-sdk-ui/src/timeline/pinned_events_loader.rs

```diff
@@ -63,20 +63,24 @@ impl PinnedEventsLoader {
             .rev()
             .collect();
 
+        // Use semaphore to limit the number of concurrent requests since RequestConfig can only
+        // set that value for the whole HttpClient
+        let semaphore = Arc::new(Semaphore::new(MAX_CONCURRENT_REQUESTS));
+
         if pinned_event_ids.is_empty() {
             return Ok(Vec::new());
         }
 
-        let request_config = Some(
-            RequestConfig::default()
-                .retry_limit(3)
-                .max_concurrent_requests(NonZeroUsize::new(MAX_CONCURRENT_REQUESTS)),
-        );
+        let request_config = Some(RequestConfig::default().retry_limit(3));
 
         let new_events = join_all(pinned_event_ids.into_iter().map(|event_id| {
+            let semaphore = semaphore.clone();
             let provider = self.room.clone();
             async move {
-                match provider.load_event(&event_id, request_config).await {
+                if let Err(err) = semaphore.acquire().await {
+                    warn!("error when loading pinned event: {err}");
+                    return None;
+                }
```

Comment:
Pretty sure this is a no-op when the semaphore could be acquired, because the guard (in the `Ok()` value) is not bound, so it's immediately dropped. Can you add a test for this? (I would even recommend making a separate PR for this feature, since it seems unrelated to the current PR's purpose, and could land separately.)

Reply:
I have no idea how to test that, to be honest. Since there is no timeout I think the execution will just get stuck there until a new permit is available.

Reply:
I replaced this with a buffered stream now. It seems to work just as fine and there is no need for the semaphore.